tagmanager/ctags/powershell.c

```diff
@@ -0,0 +1,658 @@
+
+	powershell_scope_end = strchr (token_name, *SCOPE_SEPARATOR);
+	if (powershell_scope_end)
+	{
+		int powershell_scope_len;
+		char *tag_name;
+		vString * powershell_scope = vStringNew ();
+
+		powershell_scope_len = powershell_scope_end - token_name;
+		/* cut the resulting scope string */
+		vStringNCopyS (powershell_scope, token_name, powershell_scope_len);
+		vStringTerminate (powershell_scope);
+		/* the rest after the SCOPE_SEPARATOR is the tag name */
+		tag_name = strdup (token_name + powershell_scope_len + 1);
+		/* update the token */
+		vStringClear (token->string);
+		vStringCopyS (token->string, tag_name);
+
+		access = findValidAccessType (vStringValue (powershell_scope));
+
+		free (tag_name);
+		vStringDelete (powershell_scope);
+	}
+	return access;
+}
+
+/* parse a function
+ *
+ * 	function myfunc($foo, $bar) {}
+ * 	function &myfunc($foo, $bar) {} */
+static boolean parseFunction (tokenInfo *const token)
+{
+	boolean readNext = TRUE;
+	tokenInfo *nameFree = NULL;
+	const char *access;
+
+	readToken (token);
+
+	if (token->type != TOKEN_IDENTIFIER)
+		return FALSE;
+
+	access = parsePowerShellScope (token);
+
+	nameFree = newToken ();
+	copyToken (nameFree, token, TRUE);
+	readToken (token);
+
+	if (token->type == TOKEN_OPEN_PAREN)
+	{
+		vString *arglist = vStringNew ();
+		int depth = 1;
+
+		vStringPut (arglist, '(');
+		do
+		{
+			readToken (token);
+
+			switch (token->type)
+			{
+				case TOKEN_OPEN_PAREN:  depth++; break;
+				case TOKEN_CLOSE_PAREN: depth--; break;
+				default: break;
+			}
+			/* display part */
+			switch (token->type)
+			{
+				case TOKEN_CLOSE_CURLY:		vStringPut (arglist, '}');		break;
+				case TOKEN_CLOSE_PAREN:		vStringPut (arglist, ')');		break;
+				case TOKEN_CLOSE_SQUARE:	vStringPut (arglist, ']');		break;
+				case TOKEN_COLON:			vStringPut (arglist, ':');		break;
+				case TOKEN_COMMA:			vStringCatS (arglist, ", ");	break;
+				case TOKEN_EQUAL_SIGN:		vStringCatS (arglist, " = ");	break;
+				case TOKEN_OPEN_CURLY:		vStringPut (arglist, '{');		break;
+				case TOKEN_OPEN_PAREN:		vStringPut (arglist, '(');		break;
+				case TOKEN_OPEN_SQUARE:		vStringPut (arglist, '[');		break;
+				case TOKEN_PERIOD:			vStringPut (arglist, '.');		break;
+				case TOKEN_SEMICOLON:		vStringPut (arglist, ';');		break;
+				case TOKEN_STRING:			vStringCatS (arglist, "'...'");	break;
+
+				case TOKEN_IDENTIFIER:
+				case TOKEN_KEYWORD:
+				case TOKEN_VARIABLE:
+				{
+					switch (vStringLast (arglist))
+					{
+						case 0:
+						case ' ':
+						case '{':
+						case '(':
+						case '[':
+						case '.':
+							/* no need for a space between those and the identifier */
+							break;
+
+						default:
+							vStringPut (arglist, ' ');
+							break;
+					}
+					if (token->type == TOKEN_VARIABLE)
+						vStringPut (arglist, '$');
+					vStringCat (arglist, token->string);
+					break;
+				}
+
+				default: break;
+			}
+		}
+		while (token->type != TOKEN_EOF && depth > 0);
```

Comment:
this wasn't the best design decision I ever made :D  but well, whatever, it works good enough